fuzz/fuzz_types.ml

```diff
@@ -0,0 +1,440 @@
+          (* const (AT TInt);
+             const (AT TInt32);
+             const (AT TInt64);
+             const (AT TFloat); *)
+          const (AT TString);
+          const (AT TBytes);
+          map [ at_gen ] (fun (AT t) -> AT (TList t));
+          map [ at_gen ] (fun (AT t) -> AT (TArray t));
+          map [ at_gen ] (fun (AT t) -> AT (TOption t));
+          map [ at_gen; at_gen ] (fun (AT t1) (AT t2) -> AT (TPair (t1, t2)));
+          map [ at_gen; at_gen; at_gen ] (fun (AT t1) (AT t2) (AT t3) ->
+              AT (TTriple (t1, t2, t3)));
+          map [ at_gen; at_gen ] (fun (AT t1) (AT t2) -> AT (TResult (t1, t2)));
+          map [ string; fields_gen ] (fun s fs -> AT (TRecord (s, fs)));
+          map [ string; cases_gen ] (fun s cs -> AT (TVariant (s, cs)));
+          map [ string; tags_gen ] (fun s ts -> AT (TEnum (s, ts)));
+        ])
+
+(** Convert a [t] into its [Irmin.Type] counterpart. *)
+let rec t_to_irmin : type a. a t -> a T.ty = function
+  | TUnit -> T.unit
+  | TBool -> T.bool
+  | TChar -> T.char
+  | TInt -> T.int
+  | TInt32 -> T.int32
+  | TInt64 -> T.int64
+  | TFloat -> T.float
+  | TString -> T.string
+  | TBytes -> T.bytes
+  | TList t -> T.list (t_to_irmin t)
+  | TArray t -> T.array (t_to_irmin t)
+  | TOption t -> T.option (t_to_irmin t)
+  | TPair (t1, t2) -> T.pair (t_to_irmin t1) (t_to_irmin t2)
+  | TTriple (t1, t2, t3) ->
+      T.triple (t_to_irmin t1) (t_to_irmin t2) (t_to_irmin t3)
+  | TResult (t1, t2) -> T.result (t_to_irmin t1) (t_to_irmin t2)
+  | TRecord (n, fs) -> irmin_record n fs
+  | TVariant (n, cs) -> irmin_variant n cs
+  | TEnum (n, cs) -> irmin_enum n cs
+
+(** Dynamically build an Irmin record (assuming it has 1 to 5 fields). *)
+and irmin_record : string -> field list -> dyn_record T.t = [%impl_record 5]
+
+(** Create a [dyn_record] from a list of [string * any_value] pairs. *)
+and new_dyn_record : string -> (string * any_value) list -> dyn_record =
+ fun name fields ->
+  let record = Hashtbl.create (List.length fields) in
+  List.iter (fun (k, v) -> Hashtbl.add record k v) fields;
+  (name, record)
+
+(** Create a function which retrieves a value of the given dynamic type from a
+    [dyn_record] or fails. *)
+and new_dyn_record_getter : type a. string -> string -> a t -> dyn_record -> a =
+ fun record_name field_name field_type dynrec ->
+  let curr_record_name, curr_table = dynrec in
+  if not (String.equal record_name curr_record_name) then
+    failwith "Trying to access fields from the wrong record."
+  else
+    let wrapped = Hashtbl.find curr_table field_name in
+    unwrap field_type wrapped
+
+(** Dynamically build an Irmin variant (assuming it has 1 to 5 cases). *)
+and irmin_variant : string -> case list -> dyn_variant T.t = [%impl_variant 5]
+
+(** Dynamically build an Irmin enum (assuming it has 1 to 5 cases). *)
+and irmin_enum : string -> string list -> dyn_variant T.t =
+ fun n ts -> irmin_variant n (List.map (fun t -> (t, ACT Case0)) ts)
```

Comment:
It's not clear to me why we bother distinguishing `enum` if we're not going to generate it via the `Irmin.Type.enum` combinator.

Reply:
That's true. I didn't take the time to write a specific `[%impl_enum]` because the current version of `Irmin.Type` implements enums in terms of variants anyway, but maybe your work on polymorphic calls is going to change that?

Reply:
That's unlikely to change, AFAIK. There's an argument for fuzzing `Irmin.Type.enum` since it isn't built with the other combinators (for instance, wouldn't have caught a missing `check_valid_utf8` [here](https://github.com/mirage/irmin/blob/master/src/irmin/type/type.ml#L229)).

 It shouldn't be necessary to write `[%impl_enum]`, since its combinator is fairly simple. I suggest either adding the generator or stripping out mention of `_enum` in the fuzzing.